### ISSUE TITLE
Moving operations to ram

### ISSUE THREAD

**Issue by ttk2** [Exultant/Citadel#20]
This would solve the TNT bug as well as provide a rather large performance boost if implemented. When chunks are loaded check of there are any protected blocks within that chunk and load the xyz chords and durability values of those blocks into an array list in ram. During a block break event instead of making a query to the table simply see if the arraylist contains the block just broken. This insulates players from database read writes, a TNT explosion could rack up thousands of sql actions but those can go in the back ground after the explosion, the code will just check the array list and if its protected run the decrementent function, the list of queries may take how ever long it needs to finish these operations because as far as the player is concerned they already happened. 

If this is too difficult to implement it may be abandoned, but otherwise the benefits are apparent.

**Comment by quonic** [comment]
I was thinking along the same lines, but with less specific parts of the idea. As in some buffer or queue to hold requests before any request was sent to the database.

Holding in ram would serve as the buffer, until the time is right to make changes to the database. Thus eliminating multiple database reads and writes.

**Comment by ttk2** [comment]
In the end it comes down to difficulty to implement, if [Freenor's optimizations](https://github.com/Exultant/Citadel/issues/1) for database requests prove sufficient then this may be a lot of effort for minimal gain.

**Comment by Freenor** [comment]
Just to restate the idea in simpler terms in this thread too:

The main idea is just to aggregate up input and output requests into buffers, flushing them at opportune times (or when functionally necessary) by concatenating them in RAM into a well-formed SQL query (that, say, handles a lot of things at once).  This way it opens the connection to the database as few times as possible and saves us a lot of speed.  The RAM solutions we come up with might be less than elegant, but we're dealing with serious firepower there and are more likely to see gains with mediocre RAM code over great SQL optimization (imo).

**Comment by ttk2** [comment]
Your right, ram will defiantly be faster no matter how poorly we code it. The question is will your optimizations be fast enough? And if they are fast enough then is it worth the extra time to move to ram? A performance gain from half a second to a hundredth of a second is many many times faster, but will the users even notice enough for the extra effort of a ram solution to be worth it. Thats up to whoever codes it, but a difficulty judgement is in order.

**Comment by quonic** [comment]
While trying not to over complicate things, but for future consideration.

Would a scheduler help here to reduce some of the complexity? As in it would decide what should be written to the database, in what order and even create threads to handle more requests.

Another thought: What about switching to mySQL to Posgres to increase database access performance? then we can write stored procedures and what not.

**Comment by Freenor** [comment]
ttk2 I think it's justified.  Here's why.

If we don't, someone could set off say, 100 blocks of dynamite just to troll us.  It would lock up and probably crash the server, or cause unexpected bugs.  In such a fashion, all of our effort for protecting property (say) would be for naught; someone could have fortified everything and still gotten wiped out due to some unexpected border case created by something like this.

But if we do it my way, we don't have to flush the entire buffer every round. We could limit the number of computations to some fixed, reasonable amount, and just "get to" the other blocks as we get to them.  Throttling the rate would make explosions have a "wave" which may be more or less perceptible, but it would make it so even large-scale TNT projects could proceed without fucking wrecking the server (or our mods' behaviors).

**Comment by ttk2** [comment]
Ah, that is an excellent point i wast thinking more along the lines of how it acts now. When it can get the data it needs from SQL in time the blocks are simply destroyed as if they where not protected. The way i was thinking of it optimizations would mean that the same situation would happen only with large scale numbers of tnt, for example if you had 100 TNT it would simply ignore protections, which considering that amount of tnt i would think it reasonable. But behavior like that is unpredictable, so switching to a arraylist in ram is probably a good idea, especially since arraylists have a .contains function which is probably faster than any other search available to us.

**Comment by ghost** [comment]
fixed? in d5f93dff. it now uses a reimplementation of bukkit's persistence layer (avaje ebeanserver). It uses a reimplementation because bukkit's has problems such as that it cannot be /reload-ed.
